src/order/jordan_hoelder.lean

```diff
@@ -0,0 +1,794 @@
+@[ext] lemma ext {s₁ s₂ : composition_series X} (h : ∀ x, x ∈ s₁ ↔ x ∈ s₂) : s₁ = s₂ :=
+to_list_injective $ list.eq_of_perm_of_sorted
+  (by classical; exact list.perm_of_nodup_nodup_to_finset_eq
+    s₁.to_list_nodup
+    s₂.to_list_nodup
+    (finset.ext $ by simp *))
+  s₁.to_list_sorted s₂.to_list_sorted
+
+/-- The largest element of a `composition_series` -/
+def top (s : composition_series X) : X := s (fin.last _)
+
+lemma top_mem (s : composition_series X) : s.top ∈ s :=
+mem_def.2 (set.mem_range.2 ⟨fin.last _, rfl⟩)
+
+@[simp] lemma le_top {s : composition_series X} (i : fin (s.length + 1)) : s i ≤ s.top :=
+s.strict_mono.monotone (fin.le_last _)
+
+lemma le_top_of_mem {s : composition_series X} {x : X} (hx : x ∈ s) : x ≤ s.top :=
+let ⟨i, hi⟩ := set.mem_range.2 hx in hi ▸ le_top _
+
+/-- The smallest element of a `composition_series` -/
+def bot (s : composition_series X) : X := s 0
+
+lemma bot_mem (s : composition_series X) : s.bot ∈ s :=
+mem_def.2 (set.mem_range.2 ⟨0, rfl⟩)
+
+@[simp] lemma bot_le {s : composition_series X} (i : fin (s.length + 1)) : s.bot ≤ s i :=
+s.strict_mono.monotone (fin.zero_le _)
+
+lemma bot_le_of_mem {s : composition_series X} {x : X} (hx : x ∈ s) : s.bot ≤ x :=
+let ⟨i, hi⟩ := set.mem_range.2 hx in hi ▸ bot_le _
+
+lemma length_pos_of_mem_ne {s : composition_series X}
+  {x y : X} (hx : x ∈ s) (hy : y ∈ s) (hxy : x ≠ y) :
+  0 < s.length :=
+let ⟨i, hi⟩ := hx, ⟨j, hj⟩ := hy in
+have hij : i ≠ j, from mt s.inj.2 $ λ h, hxy (hi ▸ hj ▸ h),
+hij.lt_or_lt.elim
+  (λ hij, (lt_of_le_of_lt (zero_le i)
+    (lt_of_lt_of_le hij (nat.le_of_lt_succ j.2))))
+  (λ hji, (lt_of_le_of_lt (zero_le j)
+    (lt_of_lt_of_le hji (nat.le_of_lt_succ i.2))))
+
+lemma forall_mem_eq_of_length_eq_zero {s : composition_series X}
+  (hs : s.length = 0) {x y} (hx : x ∈ s) (hy : y ∈ s) : x = y :=
+by_contradiction (λ hxy, pos_iff_ne_zero.1 (length_pos_of_mem_ne hx hy hxy) hs)
+
+/-- Remove the largest element from a `composition_series`. If the series `s`
+has length zero, then `s.erase_top = s` -/
+@[simps] def erase_top (s : composition_series X) : composition_series X :=
+{ length := s.length - 1,
+  series := λ i, s ⟨i, lt_of_lt_of_le i.2 (nat.succ_le_succ (nat.sub_le_self _ _))⟩,
+  step' := λ i, begin
+    have := s.step ⟨i, lt_of_lt_of_le i.2 (nat.sub_le_self _ _)⟩,
+    cases i,
+    exact this
+  end }
+
+lemma top_erase_top (s : composition_series X) :
+  s.erase_top.top = s ⟨s.length - 1, lt_of_le_of_lt (nat.sub_le_self _ _) (nat.lt_succ_self _)⟩ :=
+show s _ = s _, from congr_arg s
+begin
+  ext,
+  simp only [erase_top_length, fin.coe_last, fin.coe_cast_succ, fin.coe_of_nat_eq_mod,
+    fin.coe_mk, coe_coe]
+end
+
+lemma erase_top_top_le (s : composition_series X) : s.erase_top.top ≤ s.top :=
+by simp [erase_top, top, s.strict_mono.le_iff_le, fin.le_iff_coe_le_coe, nat.sub_le_self]
+
+@[simp] lemma bot_erase_top (s : composition_series X) : s.erase_top.bot = s.bot := rfl
```

Comment:
```suggestion
@[simp] lemma erase_top_bot (s : composition_series X) : s.erase_top.bot = s.bot := rfl
```

Reply:
I think my name is consistent with the convention. All of our `subgroup.map_comap_eq` etc. are named based on if they were written without dot notation.

Reply:
Ok, let's leave it like this then.